src/vulkan/engine_vulkan_debugger.cc

```diff
@@ -0,0 +1,976 @@
+    dap::VariablesResponse response;
+    request.variablesReference = variablesRef;
+    if (!Send(request, &response)) {
+      return false;
+    }
+    for (auto var : response.variables) {
+      Variable v;
+      v.name = var.name;
+      v.value = var.value;
+      if (var.variablesReference > 0) {
+        if (!GetVariables(var.variablesReference, &v.children)) {
+          return false;
+        }
+      }
+      out->emplace_back(v);
+    }
+    return true;
+  }
+
+  // GetLocals fetches the fully traversed set of local Variables from the
+  // debugger for the given stack frame.
+  // Returns true on success, false on error.
+  bool GetLocals(const dap::StackFrame& frame, Variables* out) {
+    dap::ScopesRequest scopeReq;
+    dap::ScopesResponse scopeRes;
+    scopeReq.frameId = frame.id;
+    if (!Send(scopeReq, &scopeRes)) {
+      return false;
+    }
+
+    for (auto scope : scopeRes.scopes) {
+      if (scope.presentationHint.value("") == kLocals) {
+        return GetVariables(scope.variablesReference, out);
+      }
+    }
+
+    onerror_("Locals scope not found");
+    return false;
+  }
+
+  // GetLane returns a pointer to the Variables representing the thread's SIMD
+  // lane with the given index, or nullptr if the lane was not found.
+  const Variables* GetLane(const Variables& lanes, int lane) {
+    auto out = lanes.Find(std::string(kLane) + " " + std::to_string(lane));
+    if (out == nullptr) {
+      return nullptr;
+    }
+    return &out->children;
+  }
+
+ private:
+  struct SourceCache {
+    std::unordered_map<int, SourceLines> by_ref;
+    std::unordered_map<std::string, SourceLines> by_path;
+  };
+
+  std::shared_ptr<dap::Session> session_;
+  ErrorHandler onerror_;
+  SourceCache sourceCache_;
+};
+
+// GlobalInvocationId holds a three-element unsigned integer index, used to
+// identifiy a single compute invocation.
+struct GlobalInvocationId {
+  size_t hash() const { return x << 20 | y << 10 | z; }
+  bool operator==(const GlobalInvocationId& other) const {
+    return x == other.x && y == other.y && z == other.z;
+  }
+
+  uint32_t x, y, z;
+};
+
+// InvocationKey is a tagged-union structure that identifies a single shader
+// invocation.
+struct InvocationKey {
+  // Hash is a custom hasher that can enable InvocationKeys to be used as keys
+  // in std containers.
+  struct Hash {
+    size_t operator()(const InvocationKey& key) const;
+  };
+
+  enum class Type { kGlobalInvocationId, kVertexIndex };
+  union Data {
+    GlobalInvocationId globalInvocationId;
+    uint32_t vertexId;
+  };
+
+  explicit InvocationKey(const GlobalInvocationId& id);
+  InvocationKey(Type, const Data&);
+
+  bool operator==(const InvocationKey& other) const;
+
+  // String returns a human-readable description of the key.
+  std::string String() const;
+
+  Type type;
+  Data data;
+};
+
+size_t InvocationKey::Hash::operator()(const InvocationKey& key) const {
+  size_t hash = 31 * static_cast<size_t>(key.type);
```

Comment:
Should the enum have values set explicitly? Otherwise, couldn't' the key.type end up being up to the compiler and max_int or something which will overflow the size_t when multiplied?

Reply:
I'm no c++ expert, but I believe that enums that do not have an explicit initializer are guaranteed to start at 0 and increment from there.

> "If the first enumerator has no initializer, the value of the corresponding constant is zero. An enumerator-definition without an initializer gives the enumerator the value obtained by increasing the value of the previous enumerator by one."

http://eel.is/c++draft/dcl.enum

Even if what you said were to happen, the hash would overflow, but would be still be stable.

Reply:
sgtm. I just assume the compiler will do bad things, heh